File: banking-copilot/server-python/app/token_store.py
"""
In-memory store for the Plaid access_token and a cached copy of synced
transactions. Intentionally ephemeral — persistence is a later phase. The
access_token and secret NEVER leave the server.
"""


class TokenStore:
    def __init__(self) -> None:
        self.access_token: str | None = None
        self.item_id: str | None = None
        self.cursor: str | None = None
        self.transactions: list[dict] = []

    def is_linked(self) -> bool:
        return self.access_token is not None

    def set_item(self, access_token: str, item_id: str) -> None:
        self.access_token = access_token
        self.item_id = item_id
        self.cursor = None
        self.transactions = []

    def require_token(self) -> str:
        if not self.access_token:
            raise RuntimeError(
                "No linked account. Call /api/connect first to bootstrap the "
                "Plaid sandbox Item."
            )
        return self.access_token

    def reset(self) -> None:
        self.__init__()


token_store = TokenStore()

File: banking-copilot/server/src/tokenStore.ts
/**
 * In-memory store for the Plaid access_token (and a cached copy of synced
 * transactions). This is intentionally ephemeral — storage/persistence is a
 * later phase. The access_token and secret NEVER leave the server.
 */
import type { NormalizedTxn } from "./plaid/normalize.js";

interface Store {
  accessToken: string | null;
  itemId: string | null;
  cursor: string | null;
  transactions: NormalizedTxn[];
}

const store: Store = {
  accessToken: null,
  itemId: null,
  cursor: null,
  transactions: [],
};

export const tokenStore = {
  isLinked(): boolean {
    return store.accessToken !== null;
  },
  get(): Store {
    return store;
  },
  setItem(accessToken: string, itemId: string) {
    store.accessToken = accessToken;
    store.itemId = itemId;
    store.cursor = null;
    store.transactions = [];
  },
  setCursor(cursor: string) {
    store.cursor = cursor;
  },
  setTransactions(txns: NormalizedTxn[]) {
    store.transactions = txns;
  },
  requireToken(): string {
    if (!store.accessToken) {
      throw new Error(
        "No linked account. Call the connect/init endpoint first to bootstrap the Plaid sandbox Item."
      );
    }
    return store.accessToken;
  },
  reset() {
    store.accessToken = null;
    store.itemId = null;
    store.cursor = null;
    store.transactions = [];
  },
};
